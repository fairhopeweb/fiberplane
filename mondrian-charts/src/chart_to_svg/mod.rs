mod constants;
mod create_area_path_def;
mod create_line_path_def;
mod generate_chart_content_svg;
mod generate_grid_and_axes_svg;
mod scales;
mod tick_formatters;

use crate::MondrianChart;
use constants::*;
use generate_chart_content_svg::generate_chart_content_svg;
use generate_grid_and_axes_svg::generate_grid_and_axes_svg;
use scales::Scales;

pub use self::tick_formatters::FormatterKind;

/// Options used for rendering charts.
pub struct ChartOptions<'a, S> {
    /// The width the chart should be rendered as.
    pub width: u16,

    /// The height the chart should be rendered as.
    pub height: u16,

    /// Whether an area gradient should be rendered inside area shapes and below
    /// line shapes.
    pub area_gradient_shown: bool,

    /// Whether lines should be rendered along the axes.
    pub axis_lines_shown: bool,

    /// Whether grid lines should be rendered along the ticks on the X axis.
    pub grid_columns_shown: bool,

    /// Whether grid lines should be rendered along the ticks on the Y axis.
    pub grid_rows_shown: bool,

    /// The color to use for rendering grid lines along the ticks.
    ///
    /// Must be a valid CSS color string.
    pub grid_stroke_color: &'a str,

    /// Optional dasharray to apply to the grid lines.
    pub grid_stroke_dasharray: &'a [f32],

    /// Callback to determine the color that should be used for a given shape
    /// list.
    ///
    /// Shape lists are identified by the source, of type `S`, from which they
    /// are generated and their index in the list of shape lists.
    ///
    /// The returned color must be a valid CSS color string.
    pub get_shape_list_color: &'a dyn Fn(&'a S, usize) -> &'a str,

    /// The color to use for rendering ticks along the axes.
    ///
    /// Must be a valid CSS color string.
    pub tick_color: &'a str,

    /// The type of formatter to use for ticks along the X axis.
    ///
    /// If `None`, no ticks are shown along the X axis.
    pub x_formatter: Option<FormatterKind>,

    /// The type of formatter to use for ticks along the Y axis.
    ///
    /// If `None`, no ticks are shown along the Y axis.
    pub y_formatter: Option<FormatterKind>,
}

pub fn chart_to_svg<'a, S, P>(
    chart: &'a MondrianChart<S, P>,
    options: &ChartOptions<'a, S>,
) -> String {
    let ChartOptions { width, height, .. } = options;

    let x_max = width - MARGIN_LEFT - MARGIN_RIGHT;
    let y_max = height - MARGIN_TOP - MARGIN_BOTTOM;

    let scales = Scales::new(x_max as f64, y_max as f64);

    let clip_path_id = "clip-path";
    let clip_path_y_start = -(CHART_SHAPE_OVERFLOW_MARGIN as i32);
    let clip_path_height = y_max + 2 * CHART_SHAPE_OVERFLOW_MARGIN;

    let grid_and_axes = generate_grid_and_axes_svg(chart, &scales, options);
    let chart_content = generate_chart_content_svg(chart, &scales, options);

    format!(
        "<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" width=\"{width}\" height=\"{height}\">\
            <defs>\
                <clipPath id=\"{clip_path_id}\">\
                  <rect x=\"0\" y=\"{clip_path_y_start}\" width=\"{x_max}\" height=\"{clip_path_height}\" />\
                </clipPath>\
            </defs>\
            <g transform=\"translate({MARGIN_LEFT}, {MARGIN_TOP})\">\
              {grid_and_axes}\
              <g clip-path=\"url(#{clip_path_id})\">{chart_content}</g>\
            </g>\
        </svg>"
    )
}
