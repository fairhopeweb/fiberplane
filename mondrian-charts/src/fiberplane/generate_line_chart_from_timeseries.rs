use super::utils::*;
use crate::fiberplane::{Metric, MinMax, Timeseries, TimeseriesSourceData};
use crate::types::{Axis, Line, MondrianChart, Point, Shape, ShapeList};
use std::convert::identity;

pub(crate) fn generate_line_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let buckets = create_metric_buckets(input.timeseries_data, |min_max, value| {
        min_max
            .map(|min_map: MinMax| min_map.extend_with_value(value))
            .unwrap_or_else(|| MinMax::from_value(value))
    });

    let mut x_axis = get_x_axis_from_time_range(&input.time_range);
    let mut y_axis = calculate_y_axis_range(&buckets, identity);

    for &value in input.additional_values {
        y_axis = y_axis.extend_with_value(value);
    }

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                create_shapes(&timeseries.metrics, &x_axis, &y_axis, interval)
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

fn create_shapes<'source>(
    metrics: &'source [Metric],
    x_axis: &Axis,
    y_axis: &Axis,
    interval: Option<f64>,
) -> Vec<Shape<&'source Metric>> {
    match metrics.len() {
        0 => Vec::new(),
        1 => {
            let metric = &metrics[0];
            if metric.value.is_nan() {
                Vec::new()
            } else {
                vec![Shape::Point(create_point_for_metric(
                    metric, x_axis, y_axis,
                ))]
            }
        }
        _ => {
            split_into_continuous_lines(metrics, interval)
                .into_iter()
                .map(|line| {
                    // If the line only containes one metric value, render it as a point
                    // Otherwise, render a line
                    if line.len() == 1 {
                        Shape::Point(create_point_for_metric(line[0], x_axis, y_axis))
                    } else {
                        Shape::Line(Line {
                            area_gradient_shown: None,
                            points: line
                                .into_iter()
                                .map(|metric| create_point_for_metric(metric, x_axis, y_axis))
                                .collect(),
                        })
                    }
                })
                .collect()
        }
    }
}

fn create_point_for_metric<'source>(
    metric: &'source Metric,
    x_axis: &Axis,
    y_axis: &Axis,
) -> Point<&'source Metric> {
    let time = get_time_from_timestamp(metric.time);

    Point {
        x: normalize_along_linear_axis(time, x_axis),
        y: normalize_along_linear_axis(metric.value, y_axis),
        source: metric,
    }
}
