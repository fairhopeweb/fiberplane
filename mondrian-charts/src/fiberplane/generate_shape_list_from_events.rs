use super::utils::*;
use crate::fiberplane::{PointSource, ProviderEvent, SeriesSource};
use crate::types::{Area, AreaPoint, Axis, Shape, ShapeList};

pub(crate) fn generate_shape_list_from_events<'source>(
    x_axis: &Axis,
    events: &[&'source ProviderEvent],
) -> ShapeList<SeriesSource<'source>, PointSource<'source>> {
    ShapeList {
        shapes: events
            .iter()
            .map(|event| {
                Shape::Area(Area {
                    area_gradient_shown: None,
                    points: vec![AreaPoint {
                        x: normalize_along_linear_axis(get_time_from_timestamp(event.time), x_axis),
                        y_min: 0.,
                        y_max: 1.,
                        source: PointSource::Event(event),
                    }],
                })
            })
            .collect(),
        source: SeriesSource::Events,
    }
}
