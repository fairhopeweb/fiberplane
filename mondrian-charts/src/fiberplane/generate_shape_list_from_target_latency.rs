use super::utils::*;
use crate::fiberplane::{PointSource, SeriesSource};
use crate::types::{Axis, Shape, ShapeList};
use crate::{Line, Point};

/// Generates a shape list with a single line from a fixed target latency.
pub(crate) fn generate_shape_list_from_target_latency<'source>(
    y_axis: &Axis,
    value: f64,
) -> ShapeList<SeriesSource<'source>, PointSource<'source>> {
    ShapeList {
        shapes: vec![Shape::Line(Line {
            area_gradient_shown: Some(false),
            points: vec![
                Point {
                    x: 0.,
                    y: normalize_along_linear_axis(value, y_axis),
                    source: PointSource::None,
                },
                Point {
                    x: 1.,
                    y: normalize_along_linear_axis(value, y_axis),
                    source: PointSource::None,
                },
            ],
        })],
        source: SeriesSource::TargetLatency,
    }
}
