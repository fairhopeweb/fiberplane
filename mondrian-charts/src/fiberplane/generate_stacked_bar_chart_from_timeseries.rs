use super::utils::*;
use crate::fiberplane::{Metric, StackedChartBuckets, Timeseries, TimeseriesSourceData};
use crate::types::{Axis, MondrianChart, Rectangle, Shape, ShapeList};

pub(crate) fn generate_stacked_bar_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let BucketsAndAxes {
        mut buckets,
        is_percentage,
        mut x_axis,
        y_axis,
    } = calculate_buckets_and_axes_for_stacked_chart(&input);

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        x_axis = x_axis.extend_with_interval(interval);
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let mut bar_args = BarArgs {
        bar_width: calculate_bar_width(&x_axis, interval, 1),
        buckets: &mut buckets,
        is_percentage,
        x_axis: &x_axis,
        y_axis: &y_axis,
    };

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                timeseries
                    .metrics
                    .iter()
                    .filter_map(|metric| create_bar_shape(metric, &mut bar_args))
                    .collect()
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

struct BarArgs<'axes, 'buckets> {
    bar_width: f64,
    buckets: &'buckets mut StackedChartBuckets,
    is_percentage: bool,
    x_axis: &'axes Axis,
    y_axis: &'axes Axis,
}

fn create_bar_shape<'source>(
    metric: &'source Metric,
    args: &mut BarArgs,
) -> Option<Shape<&'source Metric>> {
    if metric.value.is_nan() {
        return None;
    };

    let Some(bucket_value) = args.buckets.get_mut(&metric.time) else {
        return None;
    };

    let time = get_time_from_timestamp(metric.time);
    let value = if args.is_percentage {
        metric.value / bucket_value.total
    } else {
        metric.value
    };

    let x = normalize_along_linear_axis(time, args.x_axis) - 0.5 * args.bar_width;
    let y = bucket_value.current_y;

    let height = normalize_along_linear_axis(value, args.y_axis);
    bucket_value.current_y += height;

    Some(Shape::Rectangle(Rectangle {
        x,
        y,
        width: args.bar_width,
        height,
        source: metric,
    }))
}
