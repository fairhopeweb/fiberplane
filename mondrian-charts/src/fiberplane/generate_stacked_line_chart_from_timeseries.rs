use super::utils::*;
use crate::fiberplane::{Metric, StackedChartBuckets, Timeseries, TimeseriesSourceData};
use crate::types::{Area, AreaPoint, Axis, MondrianChart, Shape, ShapeList};

pub(crate) fn generate_stacked_line_chart_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> MondrianChart<&'source Timeseries, &'source Metric> {
    let BucketsAndAxes {
        mut buckets,
        is_percentage,
        mut x_axis,
        mut y_axis,
    } = calculate_buckets_and_axes_for_stacked_chart(&input);

    for &value in input.additional_values {
        y_axis = y_axis.extend_with_value(value);
    }

    let interval = calculate_smallest_time_interval(&buckets);
    if let Some(interval) = interval {
        attach_suggestions_to_x_axis(&mut x_axis, &buckets, interval);
    }

    let mut args = BarArgs {
        buckets: &mut buckets,
        is_percentage,
        x_axis: &x_axis,
        y_axis: &y_axis,
    };

    let shape_lists: Vec<_> = input
        .timeseries_data
        .iter()
        .map(|timeseries| ShapeList {
            shapes: if timeseries.visible {
                create_shapes(&timeseries.metrics, interval, &mut args)
            } else {
                Vec::new()
            },
            source: *timeseries,
        })
        .collect();

    MondrianChart {
        shape_lists,
        x_axis,
        y_axis,
    }
}

struct BarArgs<'axes, 'buckets> {
    buckets: &'buckets mut StackedChartBuckets,
    is_percentage: bool,
    x_axis: &'axes Axis,
    y_axis: &'axes Axis,
}

fn create_shapes<'source>(
    metrics: &'source [Metric],
    interval: Option<f64>,
    args: &mut BarArgs,
) -> Vec<Shape<&'source Metric>> {
    split_into_continuous_lines(metrics, interval)
        .iter()
        .map(|line| {
            Shape::Area(Area {
                area_gradient_shown: None,
                points: line
                    .iter()
                    .filter_map(|metric| create_point_for_metric(metric, args))
                    .collect(),
            })
        })
        .collect()
}

fn create_point_for_metric<'source>(
    metric: &'source Metric,
    args: &mut BarArgs,
) -> Option<AreaPoint<&'source Metric>> {
    let bucket_value = args.buckets.get_mut(&metric.time)?;

    let time = get_time_from_timestamp(metric.time);
    let value = if args.is_percentage {
        metric.value / bucket_value.total
    } else {
        metric.value
    };

    let y_min = bucket_value.current_y;
    let y_max = y_min + normalize_along_linear_axis(value, args.y_axis);
    bucket_value.current_y = y_max;

    Some(AreaPoint {
        x: normalize_along_linear_axis(time, args.x_axis),
        y_min,
        y_max,
        source: metric,
    })
}
