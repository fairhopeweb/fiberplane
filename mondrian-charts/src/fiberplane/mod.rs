mod constants;
mod generate_bar_chart_from_timeseries;
mod generate_line_chart_from_timeseries;
mod generate_shape_list_from_events;
mod generate_shape_list_from_target_latency;
mod generate_stacked_bar_chart_from_timeseries;
mod generate_stacked_line_chart_from_timeseries;
mod types;
mod utils;

#[cfg(test)]
mod tests;

use generate_bar_chart_from_timeseries::generate_bar_chart_from_timeseries;
use generate_line_chart_from_timeseries::generate_line_chart_from_timeseries;
use generate_shape_list_from_events::generate_shape_list_from_events;
use generate_stacked_bar_chart_from_timeseries::generate_stacked_bar_chart_from_timeseries;
use generate_stacked_line_chart_from_timeseries::generate_stacked_line_chart_from_timeseries;
pub use types::*;

use crate::MondrianChart;

use self::generate_shape_list_from_target_latency::generate_shape_list_from_target_latency;

/// Generates an abstract chart from a combination of timeseries data, events
/// and an optional target latency.
///
/// May return `None` if the graph type is unrecognized.
pub fn generate<'source>(
    input: CombinedSourceData<'source, '_>,
) -> Option<MondrianChart<SeriesSource<'source>, PointSource<'source>>> {
    let CombinedSourceData {
        graph_type,
        stacking_type,
        timeseries_data,
        events,
        target_latency,
        time_range,
    } = input;

    let chart = generate_from_timeseries(TimeseriesSourceData {
        graph_type,
        stacking_type,
        timeseries_data,
        time_range,
        additional_values: &target_latency.iter().cloned().collect::<Vec<_>>(),
    })?;

    let mut chart: MondrianChart<SeriesSource, PointSource> = chart.into();

    if graph_type == GraphType::Line && !events.is_empty() {
        chart
            .shape_lists
            .push(generate_shape_list_from_events(&chart.x_axis, events));
    }

    if let Some(latency) = target_latency {
        chart
            .shape_lists
            .push(generate_shape_list_from_target_latency(
                &chart.y_axis,
                latency,
            ));
    }

    Some(chart)
}

/// Generates an abstract chart from the given timeseries data.
///
/// May return `None` if the graph type is unrecognized.
pub fn generate_from_timeseries<'source>(
    input: TimeseriesSourceData<'source, '_, '_>,
) -> Option<MondrianChart<&'source Timeseries, &'source Metric>> {
    let abstract_chart = match (input.graph_type, input.stacking_type) {
        (GraphType::Line, StackingType::None) => generate_line_chart_from_timeseries(input),
        (GraphType::Line, _) => generate_stacked_line_chart_from_timeseries(input),
        (GraphType::Bar, StackingType::None) => generate_bar_chart_from_timeseries(input),
        (GraphType::Bar, _) => generate_stacked_bar_chart_from_timeseries(input),
        (_, _) => return None,
    };

    Some(abstract_chart)
}
