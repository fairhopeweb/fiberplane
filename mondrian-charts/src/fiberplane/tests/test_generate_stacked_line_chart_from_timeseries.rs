use crate::chart_to_svg::*;
use crate::fiberplane::{
    generate_stacked_line_chart_from_timeseries, GraphType, Metric, StackingType, TimeRange,
    Timeseries, TimeseriesSourceData, Timestamp,
};
use std::collections::BTreeMap;
use std::str::FromStr;

#[test]
fn test_generate_stacked_line_chart_from_timeseries() {
    let timeseries_a = Timeseries::builder()
        .name("dummy_data")
        .metrics(vec![
            get_metric_at_minute(0, 10.),
            get_metric_at_minute(1, 15.),
            get_metric_at_minute(2, 20.),
            get_metric_at_minute(4, 25.),
            get_metric_at_minute(5, 30.),
            get_metric_at_minute(6, 25.),
            get_metric_at_minute(7, 20.),
            get_metric_at_minute(8, 15.),
            get_metric_at_minute(9, 10.),
        ])
        .labels(BTreeMap::from([("timeseries".to_owned(), "A".to_owned())]))
        .visible(true)
        .build();

    let timeseries_b = Timeseries::builder()
        .name("dummy_data")
        .metrics(vec![
            get_metric_at_minute(0, 40.),
            get_metric_at_minute(1, 30.),
            get_metric_at_minute(2, 20.),
            get_metric_at_minute(3, 10.),
            get_metric_at_minute(4, 20.),
            get_metric_at_minute(5, 30.),
            get_metric_at_minute(6, 40.),
            get_metric_at_minute(7, 50.),
            get_metric_at_minute(8, 60.),
            get_metric_at_minute(9, 70.),
        ])
        .labels(BTreeMap::from([("timeseries".to_owned(), "B".to_owned())]))
        .visible(true)
        .build();

    let chart = generate_stacked_line_chart_from_timeseries(TimeseriesSourceData {
        graph_type: GraphType::Line,
        stacking_type: StackingType::Stacked,
        time_range: TimeRange {
            from: get_date_at_minute(0),
            to: get_date_at_minute(10),
        },
        timeseries_data: &[&timeseries_a, &timeseries_b],
        additional_values: &[],
    });

    insta::assert_yaml_snapshot!(chart);

    let svg = chart_to_svg(
        &chart,
        &ChartOptions {
            width: 640,
            height: 480,
            area_gradient_shown: true,
            axis_lines_shown: false,
            grid_columns_shown: true,
            grid_rows_shown: true,
            grid_stroke_color: "#e7e7e7",
            grid_stroke_dasharray: Default::default(),
            shape_stroke_width: None,
            get_shape_list_color: &|_, index| if index % 2 == 0 { "#c00eae" } else { "#23304a" },
            tick_color: "#a4a4a4",
            x_formatter: Some(FormatterKind::Time),
            y_formatter: Some(FormatterKind::Scientific),
        },
    );

    let svg_lines = svg
        .replace('>', ">\n")
        .split('\n')
        .filter(|line| !line.is_empty())
        .map(str::to_owned)
        .collect::<Vec<_>>();
    insta::assert_debug_snapshot!(svg_lines);
}

fn get_metric_at_minute(minute: u8, value: f64) -> Metric {
    Metric::builder()
        .time(get_date_at_minute(minute))
        .value(value)
        .build()
}

fn get_date_at_minute(minute: u8) -> Timestamp {
    Timestamp::from_str(&format!("2023-07-18T16:{minute:02}:00.000Z")).unwrap()
}
