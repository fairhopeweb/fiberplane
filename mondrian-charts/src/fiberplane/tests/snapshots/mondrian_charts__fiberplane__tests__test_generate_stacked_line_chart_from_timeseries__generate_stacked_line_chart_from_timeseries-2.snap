---
source: mondrian-charts/src/fiberplane/tests/test_generate_stacked_line_chart_from_timeseries.rs
expression: svg_lines
---
[
    "<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" width=\"640\" height=\"480\">",
    "<defs>",
    "<clipPath id=\"clip-path\">",
    "<rect x=\"0\" y=\"-4\" width=\"564\" height=\"448\" />",
    "</clipPath>",
    "</defs>",
    "<g transform=\"translate(76, 0)\">",
    "<g>",
    "<line x1=\"0\" y1=\"880.0\" x2=\"564\" y2=\"880\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"825.0\" x2=\"564\" y2=\"825\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"770.0\" x2=\"564\" y2=\"770\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"715.0\" x2=\"564\" y2=\"715\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"660.0\" x2=\"564\" y2=\"660\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"605.0\" x2=\"564\" y2=\"605\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"550.0\" x2=\"564\" y2=\"550\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"495.0\" x2=\"564\" y2=\"495\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"440.0\" x2=\"564\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"385.0\" x2=\"564\" y2=\"385\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"330.0\" x2=\"564\" y2=\"330\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"275.0\" x2=\"564\" y2=\"275\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"220.0\" x2=\"564\" y2=\"220\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"165.0\" x2=\"564\" y2=\"165\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"110.0\" x2=\"564\" y2=\"110\" stroke=\"#e7e7e7\" />",
    "<line x1=\"0\" y1=\"55.0\" x2=\"564\" y2=\"55\" stroke=\"#e7e7e7\" />",
    "</g>",
    "<line x1=\"564\" x2=\"564\" y1=\"0\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<g>",
    "<line x1=\"0.0\" y1=\"0\" x2=\"0.0\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"56.4\" y1=\"0\" x2=\"56.4\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"112.8\" y1=\"0\" x2=\"112.8\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"169.2\" y1=\"0\" x2=\"169.2\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"225.6\" y1=\"0\" x2=\"225.6\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"282.0\" y1=\"0\" x2=\"282.0\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"338.4\" y1=\"0\" x2=\"338.4\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"394.8\" y1=\"0\" x2=\"394.8\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"451.2\" y1=\"0\" x2=\"451.2\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<line x1=\"507.6\" y1=\"0\" x2=\"507.6\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "</g>",
    "<g transform=\"translate(0, 440)\">",
    "<line x1=\"0\" y1=\"0\" x2=\"564\" y2=\"0\" stroke=\"#e7e7e7\" />",
    "<text x=\"0.0\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "0:00</text>",
    "<text x=\"56.4\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "1:00</text>",
    "<text x=\"112.8\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "2:00</text>",
    "<text x=\"169.2\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "3:00</text>",
    "<text x=\"225.6\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "4:00</text>",
    "<text x=\"282.0\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "5:00</text>",
    "<text x=\"338.4\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "6:00</text>",
    "<text x=\"394.8\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "7:00</text>",
    "<text x=\"451.2\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "8:00</text>",
    "<text x=\"507.6\" y=\"20\" dy=\"16\" fill=\"#a4a4a4\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" text-anchor=\"middle\">",
    "9:00</text>",
    "</g>",
    "<g>",
    "<line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"440\" stroke=\"#e7e7e7\" />",
    "<text x=\"0\" y=\"825.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-70</text>",
    "<text x=\"0\" y=\"770.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-60</text>",
    "<text x=\"0\" y=\"715.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-50</text>",
    "<text x=\"0\" y=\"660.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-40</text>",
    "<text x=\"0\" y=\"605.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-30</text>",
    "<text x=\"0\" y=\"550.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-20</text>",
    "<text x=\"0\" y=\"495.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "-10</text>",
    "<text x=\"0\" y=\"385.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "10</text>",
    "<text x=\"0\" y=\"330.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "20</text>",
    "<text x=\"0\" y=\"275.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "30</text>",
    "<text x=\"0\" y=\"220.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "40</text>",
    "<text x=\"0\" y=\"165.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "50</text>",
    "<text x=\"0\" y=\"110.0\" dx=\"-0.45em\" dy=\"0.25em\" text-anchor=\"end\" font-family=\"sans-serif\" font-size=\"20\" font-weight=\"400\" letter-spacing=\"0\" fill=\"#a4a4a4\">",
    "60</text>",
    "</g>",
    "<g clip-path=\"url(#clip-path)\">",
    "<defs>",
    "<linearGradient id=\"line-0\" x1=\"0\" y1=\"0\" x2=\"0\" y2=\"1\">",
    "<stop offset=\"0%\" stop-color=\"#c00eae\" stop-opacity=\"0.3\" />",
    "<stop offset=\"80%\" stop-color=\"#c00eae\" stop-opacity=\"0.06\" />",
    "</linearGradient>",
    "</defs>",
    "<path d=\"M0.0,440.0L56.4,440.0L112.8,440.0L112.8,330.0L56.4,357.5L0.0,385.0Z\" stroke=\"#c00eae\" stroke-width=\"1\" fill=\"url(#line-0)\" />",
    "<defs>",
    "<linearGradient id=\"line-0\" x1=\"0\" y1=\"0\" x2=\"0\" y2=\"1\">",
    "<stop offset=\"0%\" stop-color=\"#c00eae\" stop-opacity=\"0.3\" />",
    "<stop offset=\"80%\" stop-color=\"#c00eae\" stop-opacity=\"0.06\" />",
    "</linearGradient>",
    "</defs>",
    "<path d=\"M225.6,440.0L282.0,440.0L338.4,440.0L394.8,440.0L451.2,440.0L507.6,440.0L507.6,385.0L451.2,357.5L394.8,330.0L338.4,302.5L282.0,275.0L225.6,302.5Z\" stroke=\"#c00eae\" stroke-width=\"1\" fill=\"url(#line-0)\" />",
    "<defs>",
    "<linearGradient id=\"line-1\" x1=\"0\" y1=\"0\" x2=\"0\" y2=\"1\">",
    "<stop offset=\"0%\" stop-color=\"#c00eae\" stop-opacity=\"0.3\" />",
    "<stop offset=\"80%\" stop-color=\"#c00eae\" stop-opacity=\"0.06\" />",
    "</linearGradient>",
    "</defs>",
    "<path d=\"M0.0,385.0L56.4,357.5L112.8,330.0L169.2,440.0L225.6,302.5L282.0,275.0L338.4,302.5L394.8,330.0L451.2,357.5L507.6,385.0L507.6,0.0L451.2,27.5L394.8,55.0L338.4,82.5L282.0,110.0L225.6,192.5L169.2,385.0L112.8,220.0L56.4,192.5L0.0,165.0Z\" stroke=\"#c00eae\" stroke-width=\"1\" fill=\"url(#line-1)\" />",
    "</g>",
    "</g>",
    "</svg>",
]
